fix: Add some missing TypeScript prop definititions and clean up date-picker definitions

sbusch opened the pull request:
### 🤔 This is a ...

- [x] TypeScript definition update

### 👻 What's the background?

Some props were missing from TypeScript definitions.

TypeScript definitions for date-picker Components were a bit redundant.

### 💡 Solution

Props have been added, redundancies for date-picker Components have been removed.

### 📝 Changelog

No special changelog

### ☑️ Self Check before Merge

- Doc is not needed (props are mostly standard props that are passed to the underlying HTML elements. They are most often not documented. date-picker changes doesnt need to be documented)
- Demo is not needed (same reason)
- TypeScript definition is updated (the purpose of this PR)
- Changelog is not needed

sbusch committed:
Add some missing TypeScript prop definitition, clean up date-picker definitions

zombieJ commented:
Seems conflict.

sbusch committed:
Merge branch 'master' into typescript-definitions

sbusch commented:
Conflicts resolved.

zombieJ commented on components/date-picker/interface.tsx:
```diff
@@ -31,6 +31,7 @@ export interface SinglePickerProps {
   defaultValue?: moment.Moment;
   defaultPickerValue?: moment.Moment;
   placeholder?: string;
+  renderExtraFooter?: (mode: 'date' | 'month' | 'year' | 'decade') => React.ReactNode;
```
use `DatePickerMode` instead.

sbusch replied:
@zombieJ done.

I also added `"decade"` to `DatePickerMode`, that was missing there. See commit for link to playground.

sbusch committed:
use DatePickerMode instead of redundant enumeration of modes

sbusch committed:
decade was  missing from the list of modes

See https://codesandbox.io/s/388xplj646
- activate DatePicker
- click on year in header
- click on years in header
- "decade" is printed on footer as mode for renderExtraFooter()

zombieJ commented on components/date-picker/interface.tsx:
```diff
@@ -23,21 +23,22 @@ export interface PickerProps {
   defaultPickerValue?: moment.Moment;
-  renderExtraFooter?: (mode: 'date' | 'month' | 'year' | 'decade') => React.ReactNode;
+  placeholder?: string;
+  renderExtraFooter?: (mode: DatePickerMode) => React.ReactNode;
```
`DatePickerMode | 'decade'`， `decade` is not export as mode.

sbusch replied:
I don't understand this. Do you mean that `DatePickerMode` (which is exported) should not include `'decade'`?

But `'decade'` is also a valid value for the `mode` prop ([CodeSandbox example](https://codesandbox.io/s/0q5kmwr7kl)), so it should be included in `DatePickerMode`...

zombieJ replied:
It's not in the document since we currently not export to user:
<img width="689" alt="屏幕快照 2019-05-03 下午2 14 31" src="https://user-images.githubusercontent.com/5378891/57122076-e4636600-6dad-11e9-8d47-ae34a3dcdf0f.png">

sbusch replied:
Ok, then the docs are incomplete and should be fixed:

There is definitely a "decade" mode the user can use a value for the mode prop, and `renderExtraFooter()` then gets "decade" as first argument (parameter "mode). See the CodeSandbox linked in my previous comment (<https://codesandbox.io/s/0q5kmwr7kl>), where mode prop is pinned to "decade".

You can also take a look at the more complete CodeSandbox <https://codesandbox.io/s/388xplj646>

1. activate first date picker
2. click on year in header
3. click on decade in header
4. take a look at the footer

sbusch replied:
I fixed the docs, but accidentally pushed too much into this PR branch😠

Sorry, will fix it in the next minutes.

sbusch replied:
Fixed

sbusch committed:
add decade as possible value for mode prop
